runtime/src/accounts_db.rs

```diff
@@ -6990,6 +7003,62 @@ impl AccountsDb {
                 })
                 .sum();
 
+            // subtract data.len() from accounts_data_len for all old accounts that are in the index twice
+            if pass == 0 {
+                let pubkeys_to_accounts_data_len = |pubkeys: &[Pubkey]| {
+                    let mut accounts_data_len_from_duplicates = 0;
+                    pubkeys.into_iter().for_each(|pubkey| {
+                        if let Some(entry) = self.accounts_index.get_account_read_entry(pubkey) {
+                            let slot_list = entry.slot_list();
+                            if slot_list.len() < 2 {
+                                return;
+                            }
+                            let mut slot_list = slot_list.clone();
+                            slot_list.sort_unstable_by(|a, b| a.0.cmp(&b.0));
+                            assert!(slot_list[0].0 < slot_list[1].0);
+                            slot_list
+                                .into_iter()
+                                .rev()
+                                .skip(1)
+                                .for_each(|(slot, account_info)| {
+                                    let maybe_storage_entry = self
+                                        .storage
+                                        .get_account_storage_entry(slot, account_info.store_id);
+                                    let mut accessor = LoadedAccountAccessor::Stored(
+                                        maybe_storage_entry
+                                            .map(|entry| (entry, account_info.offset)),
+                                    );
+                                    let loaded_account = accessor.check_and_get_loaded_account();
+                                    let account = loaded_account.take_account();
+                                    accounts_data_len_from_duplicates += account.data().len();
+                                });
+                        }
+                    });
+                    accounts_data_len_from_duplicates as u64
+                };
```

Comment:
And this is the closure to get the accounts data len from the pubkeys chunk. Again, no atomic operations in here, yay!